docs/design_notes/module_lifecycle.rst

```diff
@@ -0,0 +1,378 @@
+The Module lifecycle
+######################
+
+.. testsetup::
+
+  from typing import Any, Callable, Iterable
+  import flax
+  from flax import linen as nn
+  from jax import random
+  import jax
+
+
+This design note is intended for users who are already familiar with linen Modules but want to understand more about the design principles behind the abstraction. This note should give you a good understanding of the assumptions and guarantees the Module API is built upon. If you have no practical experience with Modules yet, check out the `MNIST Tutorial <https://flax.readthedocs.io/en/latest/notebooks/annotated_mnist.html>`_.
+
+linen Modules offer a pythonic abstracton on top of Flax core. The ``Module`` abstraction allows you to create classes that have state, parameters and randomness on top of JAX. This is a practical guide to the design and behavior of the ``Module`` class. By the end, you should feel comfortable to go off the beaten track and use Modules in new ways.
+
+
+Overview
+***********
+
+Definition
+=============
+
+Let's start with a high-level overview of the Module lifecycle. We start by defining a simple Module:
+
+
+.. testcode::
+
+  class MLP(nn.Module):
+    # 1. Attribute annotations
+    hidden_size: int
+    out_size: int
+
+    # 2. The ``setup`` method
+    def setup(self):
+      self.hidden = nn.Dense(self.hidden_size)
+      self.out = nn.Dense(self.out_size)
+
+    # 3. User methods  
+    def __call__(self, x):
+      a = self.hidden(x)
+      h = nn.relu(a)
+      return self.out(h)
+
+
+This Module consists of:
+
+#. **Attribute annotations**, defined as `dataclass <https://docs.python.org/3/library/dataclasses.html>`_ fields. These annotations automatically define a constructor.
+#. **The ``setup`` method**, which creates submodules and assigns them to attributes.
+#. **User methods**. By convention, most Modules have just one  ``__call__`` method, but you can define multiple methods or use different method names.
+
+Construction/initialization
+=============================
+
+Now we want to construct and use the ``MLP`` Module:
+
+
+.. testcode::
+
+  mlp = MLP(hidden_size=5, out_size=3)
+  x = jax.numpy.ones((1, 2))
+  variables = mlp.init(random.PRNGKey(0), x)
+  y = mlp.apply(variables, x)
+
+
+First, we construct an instance of ``MLP`` and pass the construction attributes. Note that construction here is different from what you might expect if you are not used to Functional Programming patterns. The ``MLP`` constructor does not actually create variables or any internal state whatsoever. It's best to think of it as a specification or template of the Module that contains functionality but no data.
+
+Let's first take a closer look at initialization. Surprisingly, there actually is no special initialization path in Flax. Calling ``init`` is just a special case of ``apply`` and we could also code it as:
```

Comment:
Nit: to avoid repeating "first" in 2 consecutive paragraphs, maybe try this 

```suggestion
Let's take a closer look at initialization. Surprisingly, there actually is no special initialization path in Flax. Calling ``init`` is just a special case of ``apply``, which you can also write as:
```

And maybe try rephrasing the first or second "special" to avoid repeating "special" twice.